File: android-app/app/src/main/kotlin/io/github/kpganon/termuxmpvcontrols/Protocol.kt
package io.github.kpganon.termuxmpvcontrols

import org.json.JSONObject

/** Client half of the wire format documented in `docs/protocol.md`. */
object Protocol {

    const val VERSION = 1

    sealed interface Art {
        /** A remote image, with a second URL to try when the first 404s. */
        data class Remote(val url: String, val fallbackUrl: String?) : Art

        /** Bytes inlined by the daemon, for local files with a sidecar cover. */
        data class Inline(val base64: String, val mime: String) : Art
    }

    data class State(
        val playing: Boolean = false,
        val title: String = "",
        val artist: String? = null,
        val album: String? = null,
        val index: Int = 0,
        val count: Int = 0,
        val positionMs: Long = 0L,
        val durationMs: Long? = null,
        val art: Art? = null,
        val idle: Boolean = true,
        val url: String? = null,
    )

    data class PlaylistEntry(
        val index: Int,
        val title: String,
        val current: Boolean,
        val url: String? = null,
        /** A deleted, private or region-blocked video: mpv never resolved a title for it. */
        val unavailable: Boolean = false,
    )

    data class Refresh(
        val status: String,
        val added: Int = 0,
        val total: Int = 0,
        val reason: String? = null,
    ) {
        val isRunning: Boolean get() = status == "running"
    }

    sealed interface Message {
        data class Hello(val version: Int, val mpv: String?) : Message
        data class StateChanged(val state: State) : Message
        data class PlaylistChanged(val entries: List<PlaylistEntry>) : Message
        data object Bye : Message
        data class Error(val reason: String) : Message
        data class RefreshProgress(val refresh: Refresh) : Message
        data class Unknown(val type: String?) : Message
    }

    fun parse(line: String): Message {
        val json = JSONObject(line)
        return when (val type = json.optString("type")) {
            "hello" -> Message.Hello(
                version = json.optInt("version", VERSION),
                mpv = json.optStringOrNull("mpv"),
            )

            "state" -> Message.StateChanged(parseState(json))
            "playlist" -> Message.PlaylistChanged(parsePlaylist(json))
            "bye" -> Message.Bye
            "refresh" -> Message.RefreshProgress(
                Refresh(
                    status = json.optString("status", "error"),
                    added = json.optInt("added", 0),
                    total = json.optInt("total", 0),
                    reason = json.optStringOrNull("reason"),
                )
            )
            "error" -> Message.Error(json.optString("reason", "unknown"))
            else -> Message.Unknown(type)
        }
    }

    private fun parseState(json: JSONObject): State = State(
        playing = json.optBoolean("playing", false),
        title = json.optString("title", ""),
        artist = json.optStringOrNull("artist"),
        album = json.optStringOrNull("album"),
        index = json.optInt("index", 0),
        count = json.optInt("count", 0),
        positionMs = (json.optDouble("position", 0.0) * 1000).toLong(),
        durationMs = json.optDoubleOrNull("duration")?.let { (it * 1000).toLong() },
        art = json.optJSONObject("art")?.let(::parseArt),
        idle = json.optBoolean("idle", true),
        url = json.optStringOrNull("url"),
    )

    private fun parseArt(json: JSONObject): Art? {
        json.optStringOrNull("data")?.let { data ->
            return Art.Inline(base64 = data, mime = json.optString("mime", "image/jpeg"))
        }
        json.optStringOrNull("url")?.let { url ->
            return Art.Remote(url = url, fallbackUrl = json.optStringOrNull("fallbackUrl"))
        }
        return null
    }

    private fun parsePlaylist(json: JSONObject): List<PlaylistEntry> {
        val array = json.optJSONArray("entries") ?: return emptyList()
        return (0 until array.length()).mapNotNull { i ->
            val entry = array.optJSONObject(i) ?: return@mapNotNull null
            PlaylistEntry(
                index = entry.optInt("index", i),
                title = entry.optString("title", ""),
                current = entry.optBoolean("current", false),
                url = entry.optStringOrNull("url"),
                unavailable = entry.optBoolean("unavailable", false),
            )
        }
    }

    fun hello(token: String?): String =
        JSONObject().put("type", "hello").put("version", VERSION).apply {
            if (!token.isNullOrEmpty()) put("token", token)
        }.toString()

    fun command(name: String): String =
        JSONObject().put("type", "cmd").put("name", name).toString()

    fun seek(positionMs: Long): String =
        JSONObject().put("type", "cmd").put("name", "seek")
            .put("position", positionMs / 1000.0).toString()

    fun goto(index: Int): String =
        JSONObject().put("type", "cmd").put("name", "goto").put("index", index).toString()

    fun playUrl(url: String): String =
        JSONObject().put("type", "cmd").put("name", "play-url").put("url", url).toString()

    fun refreshPlaylist(): String =
        JSONObject().put("type", "cmd").put("name", "refresh-playlist").toString()

    private fun JSONObject.optStringOrNull(key: String): String? {
        if (!has(key) || isNull(key)) return null
        return optString(key).takeIf { it.isNotEmpty() }
    }

    private fun JSONObject.optDoubleOrNull(key: String): Double? {
        if (!has(key) || isNull(key)) return null
        val value = optDouble(key)
        return if (value.isNaN()) null else value
    }
}

File: daemon/src/mpvbridge/protocol.py
"""Wire format shared with the Android companion app.

Newline-delimited JSON in both directions over a loopback TCP socket. Every message is a JSON
object with a ``type`` key. See ``docs/protocol.md`` for the full contract.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

PROTOCOL_VERSION = 1
DEFAULT_PORT = 7355
DEFAULT_HOST = "127.0.0.1"

# server -> client
SERVER_HELLO = "hello"
SERVER_STATE = "state"
SERVER_PLAYLIST = "playlist"
SERVER_BYE = "bye"
SERVER_REFRESH = "refresh"
SERVER_LIBRARY = "library"
SERVER_ERROR = "error"

# client -> server
CLIENT_HELLO = "hello"
CLIENT_COMMAND = "cmd"

# command names a client may send
CMD_PLAY = "play"
CMD_PAUSE = "pause"
CMD_TOGGLE = "toggle"
CMD_NEXT = "next"
CMD_PREVIOUS = "previous"
CMD_STOP = "stop"
CMD_SEEK = "seek"
CMD_GOTO = "goto"
CMD_REFRESH = "refresh"
CMD_PLAY_URL = "play-url"
CMD_REFRESH_PLAYLIST = "refresh-playlist"
CMD_SHUFFLE = "shuffle"
CMD_UNSHUFFLE = "unshuffle"
CMD_LIBRARY = "library"
CMD_ADD_PLAYLIST = "add-playlist"
CMD_LOAD_PLAYLIST = "load-playlist"
CMD_REMOVE_PLAYLIST = "remove-playlist"

COMMAND_NAMES = frozenset(
    {
        CMD_PLAY,
        CMD_PAUSE,
        CMD_TOGGLE,
        CMD_NEXT,
        CMD_PREVIOUS,
        CMD_STOP,
        CMD_SEEK,
        CMD_GOTO,
        CMD_REFRESH,
        CMD_PLAY_URL,
        CMD_REFRESH_PLAYLIST,
        CMD_SHUFFLE,
        CMD_UNSHUFFLE,
        CMD_LIBRARY,
        CMD_ADD_PLAYLIST,
        CMD_LOAD_PLAYLIST,
        CMD_REMOVE_PLAYLIST,
    }
)


@dataclass(slots=True)
class PlayerState:
    """The snapshot the app renders. Field names are the wire names."""

    playing: bool = False
    title: str = ""
    artist: str | None = None
    album: str | None = None
    index: int = 0
    count: int = 0
    position: float = 0.0
    duration: float | None = None
    art: dict[str, str] | None = None
    idle: bool = True
    url: str | None = None
    #: The playlist these entries came from, so the app can name what is playing. ``None`` when
    #: mpv was handed loose files rather than a playlist.
    source_url: str | None = None
    source_title: str | None = None

    def to_message(self) -> dict[str, Any]:
        return {
            "type": SERVER_STATE,
            "playing": self.playing,
            "title": self.title,
            "artist": self.artist,
            "album": self.album,
            "index": self.index,
            "count": self.count,
            "position": round(self.position, 3),
            "duration": round(self.duration, 3) if self.duration is not None else None,
            "art": self.art,
            "idle": self.idle,
            "url": self.url,
            "source": (
                {"url": self.source_url, "title": self.source_title}
                if self.source_url
                else None
            ),
        }


@dataclass(slots=True)
class PlaylistEntry:
    index: int
    title: str
    current: bool = False
    url: str | None = None
    #: True when mpv gave us no real title -- a deleted, private or region-blocked video. The app
    #: hides these by default.
    unavailable: bool = False

    def to_message(self) -> dict[str, Any]:
        return {
            "index": self.index,
            "title": self.title,
            "current": self.current,
            "url": self.url,
            "unavailable": self.unavailable,
        }


@dataclass(slots=True)
class Playlist:
    entries: list[PlaylistEntry] = field(default_factory=list)

    def to_message(self) -> dict[str, Any]:
        return {
            "type": SERVER_PLAYLIST,
            "entries": [entry.to_message() for entry in self.entries],
        }


@dataclass(slots=True)
class SavedPlaylist:
    """One playlist the daemon has cached, and can start playing without re-resolving it."""

    url: str
    title: str | None = None
    count: int = 0
    fetched: int = 0
    current: bool = False

    def to_message(self) -> dict[str, Any]:
        return {
            "url": self.url,
            "title": self.title,
            "count": self.count,
            "fetched": self.fetched,
            "current": self.current,
        }


def library_message(playlists: list[SavedPlaylist]) -> dict[str, Any]:
    return {
        "type": SERVER_LIBRARY,
        "playlists": [playlist.to_message() for playlist in playlists],
    }


def hello_message(mpv_version: str | None) -> dict[str, Any]:
    return {
        "type": SERVER_HELLO,
        "version": PROTOCOL_VERSION,
        "mpv": mpv_version,
    }


def error_message(reason: str) -> dict[str, Any]:
    return {"type": SERVER_ERROR, "reason": reason}


#: Which job a ``refresh`` message is reporting on. They share a message type because from the
#: app's side they are the same wait, but "added 855 tracks" and "found 855 tracks" are not the
#: same sentence, so the app has to be able to tell them apart.
REFRESH_KIND_RECHECK = "refresh"
REFRESH_KIND_ADD = "add"
REFRESH_KIND_LOAD = "load"


def refresh_message(
    status: str,
    added: int = 0,
    total: int = 0,
    reason: str | None = None,
    title: str | None = None,
    url: str | None = None,
    kind: str = REFRESH_KIND_RECHECK,
) -> dict[str, Any]:
    """Progress for a playlist fetch; yt-dlp on a large playlist takes many seconds.

    Re-checking the current playlist, adding a new one and loading an unseen one all report
    through this. ``title`` and ``url`` name the playlist; ``kind`` names the job.
    """
    return {
        "type": SERVER_REFRESH,
        "status": status,
        "added": added,
        "total": total,
        "reason": reason,
        "title": title,
        "url": url,
        "kind": kind,
    }
